src/particle.cpp

```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
+
+      // Adjust the particle's location and direction.
+      Position r {coord[0].xyz};
+      Direction u {coord[0].uvw};
+      bool rotational = other->periodic_translate(surf_p, r, u);
+      std::copy(&r.x, &r.x + 3, coord[0].xyz);
+      std::copy(&u.x, &u.x + 3, coord[0].uvw);
+
+      // Reassign particle's surface
+      // TODO: off-by-one
+      surface = rotational ?
+        surf_p->i_periodic_ + 1 :
+        std::copysign(surf_p->i_periodic_ + 1, surface);
+    }
+
+    // Figure out what cell particle is in now
+    n_coord = 1;
+
+    if (!find_cell(this, true)) {
+      this->mark_as_lost("Couldn't find particle after hitting periodic "
+        "boundary on surface " + std::to_string(surf->id_) + ".");
+      return;
+    }
+
+    // Set previous coordinate going slightly past surface crossing
+    last_xyz_current[0] = coord[0].xyz[0] + TINY_BIT * coord[0].uvw[0];
+    last_xyz_current[1] = coord[0].xyz[1] + TINY_BIT * coord[0].uvw[1];
+    last_xyz_current[2] = coord[0].xyz[2] + TINY_BIT * coord[0].uvw[2];
+
+    // Diagnostic message
+    if (settings::verbosity >= 10 || simulation::trace) {
+      write_message("    Hit periodic boundary on surface " +
+        std::to_string(surf->id_));
+    }
+    return;
+  }
+
+  // ==========================================================================
+  // SEARCH NEIGHBOR LISTS FOR NEXT CELL
+
+#ifdef DAGMC
+  if (settings::dagmc) {
+    auto cellp = dynamic_cast<DAGCell*>(model::cells[last_cell[0]]);
+    // TODO: off-by-one
+    auto surfp = dynamic_cast<DAGSurface*>(model::surfaces[std::abs(surface) - 1]);
+    int32_t i_cell = next_cell(cellp, surfp) - 1;
+    // save material and temp
+    last_material = material;
+    last_sqrtkT = sqrtkT;
+    // set new cell value
+    coord[0].cell = i_cell;
+    cell_instance = 0;
+    // TODO: off-by-one
+    int mat = model::cells[i_cell]->material_[0];
+    material = (mat == MATERIAL_VOID) ? mat : mat + 1;
+    sqrtkT = model::cells[i_cell]->sqrtkT_[0];
+    return;
+  }
+#endif
+
+  if (find_cell(this, true)) return;
+
+  // ==========================================================================
+  // COULDN'T FIND PARTICLE IN NEIGHBORING CELLS, SEARCH ALL CELLS
+
+  // Remove lower coordinate levels and assignment of surface
+  surface = ERROR_INT;
+  n_coord = 1;
+  bool found = find_cell(this, false);
+
+  if (settings::run_mode != RUN_MODE_PLOTTING && (!found)) {
+    // If a cell is still not found, there are two possible causes: 1) there is
+    // a void in the model, and 2) the particle hit a surface at a tangent. If
+    // the particle is really traveling tangent to a surface, if we move it
+    // forward a tiny bit it should fix the problem.
+
+    n_coord = 1;
+    coord[0].xyz[0] += TINY_BIT * coord[0].uvw[0];
+    coord[0].xyz[1] += TINY_BIT * coord[0].uvw[1];
+    coord[0].xyz[2] += TINY_BIT * coord[0].uvw[2];
+
+    // Couldn't find next cell anywhere// This probably means there is an actual
```

Comment:
Next time I'd suggest `^\s\+!` for the REGEX :p

Reply:
I'm not that sophisticated ;)